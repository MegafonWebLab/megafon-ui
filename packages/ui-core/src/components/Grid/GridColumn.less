@import '~styles/base.less';

.mfui-grid-column {
    box-sizing: border-box;
    &_grow {
        flex-grow: 1;
    }
    &_align_right {
        margin-left: auto;
    }
    &_align_left {
        margin-right: auto;
    }
    &_align_center {
        margin-right: auto;
        margin-left: auto;
    }

    &_flex {
        display: flex;
        flex-shrink: 0;
        flex-grow: 0;
    }

    .generateColumns(all);
    .generateOffsets(all);
    .generateOrder(all-order);

    @media @desktopMU {
        .generateColumns(wide);
        .generateOrder(wide-order);
        .generateOffsets(wide);
    }
    @media @desktopS {
        .generateColumns(desktop);
        .generateOrder(desktop-order);
        .generateOffsets(desktop);
    }
    @media @mobileB {
        .generateColumns(tablet);
        .generateOrder(tablet-order);
        .generateOffsets(tablet);
    }
    @media @mobileSM {
        .generateColumns(mobile);
        .generateOrder(mobile-order);
        .generateOffsets(mobile);
    }
}

.generateColumns(@resolution, @columnCount: 12, @i: 1) when (@i <= @columnCount) {
    &_@{resolution}_@{i} {
        width: (100% / @columnCount) * @i;
    }
    .generateColumns(@resolution, @columnCount, (@i + 1));
}

.generateOrder(@class, @maxOrder:100, @z:0) when (@z <= @maxOrder) {
    &_@{class}_@{z} {
        order: @z;
    }
    .generateOrder(@class, @maxOrder, (@z + 1));
}

.generateOffsets(@resolution, @columnCount: 12, @i: 1) when (@i <= @columnCount) {
    &_offset-@{resolution}_@{i} {
        margin-left: (100% / @columnCount) * @i;
    }
    .generateOffsets(@resolution, @columnCount, (@i + 1));
}
