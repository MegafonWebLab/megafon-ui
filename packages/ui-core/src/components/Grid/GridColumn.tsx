import * as React from 'react';
import * as PropTypes from 'prop-types';
import './GridColumn.less';
import cnCreate from 'utils/cnCreate';

const GridSizeValues = PropTypes.oneOf(['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12']);
export type TGridSizeValues = '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' | '10' | '11' | '12';

export interface IGridColumn {
    /** Size of columns on wide screens */
    wide?: TGridSizeValues;
    /** Size of columns on desktop screens */
    desktop?: TGridSizeValues;
    /** Size of columns on tablet screens */
    tablet?: TGridSizeValues;
    /** Size of columns on mobile screens */
    mobile?: TGridSizeValues;
    /** Size of columns on all screens */
    all?: TGridSizeValues;

    /** Order on all screens */
    orderAll?: string;
    /** Order on wide screens */
    orderWide?: string;
    /** Order on desktop screens */
    orderDesktop?: string;
    /** Order on tablet screens */
    orderTablet?: string;
    /** Order on mobile screens */
    orderMobile?: string;

    /** Size of offset on wide screens */
    offsetWide?: TGridSizeValues;
    /** Size of offset on desktop screens */
    offsetDesktop?: TGridSizeValues;
    /** Size of offset on tablet screens */
    offsetTablet?: TGridSizeValues;
    /** Size of offset on mobile screens */
    offsetMobile?: TGridSizeValues;
    /** Size of offset on all screens */
    offsetAll?: TGridSizeValues;

    /** Custom alignment of column */
    align?: 'right' | 'left' | 'center';
    /** Column flex grow */
    grow?: boolean;
    /** Column as flex container */
    flex?: boolean;
    /** Custom class name */
    className?: string;
    children: JSX.Element[] | Element[] | JSX.Element | string | Element;
}

const cn = cnCreate('mfui-grid-column');
class GridColumn extends React.Component<IGridColumn, {}> {
    static propTypes = {
        wide: GridSizeValues,
        desktop: GridSizeValues,
        tablet: GridSizeValues,
        mobile: GridSizeValues,
        all: GridSizeValues,

        orderAll: PropTypes.string,
        orderWide: PropTypes.string,
        orderDesktop: PropTypes.string,
        orderTablet: PropTypes.string,
        orderMobile: PropTypes.string,

        offsetWide: GridSizeValues,
        offsetDesktop: GridSizeValues,
        offsetTablet: GridSizeValues,
        offsetMobile: GridSizeValues,
        offsetAll: GridSizeValues,

        align: PropTypes.oneOf(['right', 'left', 'center']),
        grow: PropTypes.bool,
        flex: PropTypes.bool,
        className: PropTypes.string,
        children: PropTypes.oneOfType([
            PropTypes.arrayOf(PropTypes.element),
            PropTypes.element,
            PropTypes.string,
        ]),
    };

    static defaultProps = {
        all: '12',
        flex: false,
        grow: false,
        orderWide: '0',
        orderDesktop: '0',
        orderTablet: '0',
        orderMobile: '0',
        orderAll: '0',
    };

    render() {
        const {
            all,
            wide,
            desktop,
            tablet,
            mobile,
            orderAll,
            orderWide,
            orderDesktop,
            orderTablet,
            orderMobile,
            offsetAll,
            offsetWide,
            offsetDesktop,
            offsetTablet,
            offsetMobile,
            className,
            flex,
            grow,
            align,
            children,
        } = this.props;

        return (
            <div
                className={cn(
                    '',
                    {
                        flex,
                        grow,
                        align,
                        all,
                        wide,
                        desktop,
                        tablet,
                        mobile,
                        'all-order': orderAll,
                        'wide-order': orderWide,
                        'desktop-order': orderDesktop,
                        'tablet-order': orderTablet,
                        'mobile-order': orderMobile,
                        'offset-all': offsetAll,
                        'offset-wide': offsetWide,
                        'offset-desktop': offsetDesktop,
                        'offset-tablet': offsetTablet,
                        'offset-mobile': offsetMobile,
                    },
                    className
                )}
            >
                {children}
            </div>
        );
    }
}

export default GridColumn;
