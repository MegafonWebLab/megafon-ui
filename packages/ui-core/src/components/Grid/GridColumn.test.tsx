import * as React from 'react';
import { shallow } from 'enzyme';
import GridColumn from './GridColumn';

describe('<GridColumn />', () => {
    it('it renders GridColumn without props', () => {
        const wrapper = shallow(<GridColumn>item</GridColumn>);
        expect(wrapper).toMatchSnapshot();
    });

    it('it renders GridColumn with props on all screens', () => {
        const wrapper = shallow(
            <GridColumn
                wide="8"
                desktop="4"
                tablet="3"
                mobile="1"
                all="10"
                align="right"
                orderAll="2"
                offsetAll="4"
                grow
                flex
            >
                item
            </GridColumn>
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('it renders GridColumn with props on different screens', () => {
        const wrapper = shallow(
            <GridColumn
                className="test-class"
                wide="8"
                desktop="4"
                tablet="3"
                mobile="1"
                all="10"
                align="right"
                orderWide="2"
                orderDesktop="1"
                orderTablet="3"
                orderMobile="4"
                offsetWide="8"
                offsetDesktop="7"
                offsetTablet="6"
                offsetMobile="5"
                grow
                flex
            >
                item
            </GridColumn>
        );
        expect(wrapper).toMatchSnapshot();
    });
});
