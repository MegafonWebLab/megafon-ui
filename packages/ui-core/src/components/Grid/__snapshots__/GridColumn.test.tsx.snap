// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GridColumn /> it renders GridColumn with props on all screens 1`] = `
<div
  className="mfui-grid-column mfui-grid-column_flex mfui-grid-column_grow mfui-grid-column_align_right mfui-grid-column_all_10 mfui-grid-column_wide_8 mfui-grid-column_desktop_4 mfui-grid-column_tablet_3 mfui-grid-column_mobile_1 mfui-grid-column_all-order_2 mfui-grid-column_wide-order_0 mfui-grid-column_desktop-order_0 mfui-grid-column_tablet-order_0 mfui-grid-column_mobile-order_0 mfui-grid-column_offset-all_4"
>
  item
</div>
`;

exports[`<GridColumn /> it renders GridColumn with props on different screens 1`] = `
<div
  className="mfui-grid-column mfui-grid-column_flex mfui-grid-column_grow mfui-grid-column_align_right mfui-grid-column_all_10 mfui-grid-column_wide_8 mfui-grid-column_desktop_4 mfui-grid-column_tablet_3 mfui-grid-column_mobile_1 mfui-grid-column_all-order_0 mfui-grid-column_wide-order_2 mfui-grid-column_desktop-order_1 mfui-grid-column_tablet-order_3 mfui-grid-column_mobile-order_4 mfui-grid-column_offset-wide_8 mfui-grid-column_offset-desktop_7 mfui-grid-column_offset-tablet_6 mfui-grid-column_offset-mobile_5 test-class"
>
  item
</div>
`;

exports[`<GridColumn /> it renders GridColumn without props 1`] = `
<div
  className="mfui-grid-column mfui-grid-column_all_12 mfui-grid-column_all-order_0 mfui-grid-column_wide-order_0 mfui-grid-column_desktop-order_0 mfui-grid-column_tablet-order_0 mfui-grid-column_mobile-order_0"
>
  item
</div>
`;
